torch/csrc/jit/passes/autocast.cpp

```diff
@@ -0,0 +1,456 @@
+  for (auto input : node->inputs()) {
+    if (auto tensor_type = input->type()->cast<TensorType>()) {
+      const auto dtype = tensor_type->scalarType();
+      if (!dtype.has_value() || *dtype == at::ScalarType::Float) {
+        castTensorInputs(node, aten::_autocast_to_full_precision, context);
+        return;
+      }
+    }
+  }
+}
+
+void handleBlock(Block* block, AutocastContext initial_state) {
+  std::stack<AutocastScope> autocast_stack;
+
+  c10::optional<bool> incompatible_amp = c10::nullopt;
+
+  // The current autocast enabled/disabled state
+  auto current_state = [&] {
+    return autocast_stack.empty() ? initial_state
+                                  : autocast_stack.top().context;
+  };
+
+  for (Node* node : block->nodes()) {
+    switch (node->kind()) {
+      case prim::CallFunction:
+        // TODO: limit it only to amp related node;
+        TORCH_INTERNAL_ASSERT(
+            !incompatible_amp.has_value() || incompatible_amp.value(),
+            "Calls are not expected with AMP & JIT");
+        incompatible_amp = true;
+        break;
+
+      case prim::CallMethod:
+        // TODO: limit it only to amp related node;
+        if (auto class_type = node->input(0)->type()->cast<ClassType>()) {
+          const auto& name = node->s(attr::name);
+          const auto& function = class_type->getMethod(name);
+          if (!function.isGraphFunction()) {
+            TORCH_INTERNAL_ASSERT(
+                !incompatible_amp.has_value() || incompatible_amp.value(),
+                "Calls are not expected with AMP & JIT");
+            incompatible_amp = true;
+          }
+        } else {
+          TORCH_INTERNAL_ASSERT(
+              !incompatible_amp.has_value() || incompatible_amp.value(),
+              "Unexpected prim::CallMethod form with AMP & JIT");
+          incompatible_amp = true;
+        }
+        break;
+
+      case prim::Enter:
+        if (auto autocast_scope =
+                parseAutocast(node->input(), current_state())) {
+          if (node->hasUses()) {
+            // TODO: better error message
+            AT_ERROR("`with autocast() as ...` is not supported");
+          }
+          TORCH_INTERNAL_ASSERT(
+              !incompatible_amp.has_value() || !incompatible_amp.value(),
+              "Unsupported case by AMP & JIT");
+          incompatible_amp = false;
+          autocast_stack.push(*autocast_scope);
+        }
+        break;
+
+      case prim::Exit:
+        if (isAutocastNode(node->input(0))) {
+          TORCH_INTERNAL_ASSERT(!autocast_stack.empty());
+          TORCH_INTERNAL_ASSERT(autocast_stack.top().instance == node->input());
+          TORCH_INTERNAL_ASSERT(
+              !incompatible_amp.has_value() || !incompatible_amp.value(),
+              "Unsupported case by AMP & JIT");
+          incompatible_amp = false;
+          autocast_stack.pop();
+        }
+        break;
+
+      // CastPolicy::fp16 (cast all inputs to float16)
+      case aten::_convolution:
```

Comment:
could you add a link to the corresponding eager logic

Reply:
Good catch. Will do.

This is a pretty temporary thing, I was chatting with Carilli that we should have used the logic from eager with some added exceptions. I'll have a TODO here as well.